File: solution/2187. Minimum Time to Complete Trips/2187.py
class Solution(object):
    def minimumTime(self, time, totalTrips):
        """
        :type time: List[int]
        :type totalTrips: int
        :rtype: int
        """
        def count(t):
            res = 0
            for x in time:
                res = res + (t // x)
            return res
        
        l, r = 0, sys.maxint
        while l < r:
            mid = (l + r) // 2
            trips = count(mid)
            if trips < totalTrips:
                l = mid + 1
            else:
                r = mid
        return l

File: solution/2187. Minimum Time to Complete Trips/2187.cpp
class Solution {
public:
    long long minimumTime(vector<int>& time, int totalTrips) {
        long l = 0, r = (long)*min_element(time.begin(), time.end()) * totalTrips;
        while (l < r) {
            long mid = l + (r - l) / 2;
            long trips = count(time, mid);
            if (trips < totalTrips)
                l = mid + 1;
            else
                r = mid;
        }
        return l;
    }
    
    long count(vector<int>& time, long t) {
        long trips = 0;
        for (int x : time) {
            trips += t / x;
        }
        return trips;
    }
};
